ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/DebugDataDumper.java

```diff
@@ -0,0 +1,186 @@
+
+  private static final String GOSSIP_MESSAGES_DIR = "gossip_messages";
+  private static final String DECODING_ERROR_SUB_DIR = "decoding_error";
+  private static final String REJECTED_SUB_DIR = "rejected";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  private boolean enabled;
+  private final Path directory;
+
+  public DebugDataDumper(final Path directory, final boolean enabled) {
+    this.enabled = enabled;
+    this.directory = directory;
+    if (!enabled) {
+      return;
+    }
+
+    final Path gossipMessagesPath = this.directory.resolve(GOSSIP_MESSAGES_DIR);
+    createDirectory(gossipMessagesPath, GOSSIP_MESSAGES_DIR, "gossip messages");
+    createDirectory(
+        gossipMessagesPath.resolve(DECODING_ERROR_SUB_DIR),
+        DECODING_ERROR_SUB_DIR,
+        "gossip messages with decoding errors");
+    createDirectory(
+        gossipMessagesPath.resolve(REJECTED_SUB_DIR), REJECTED_SUB_DIR, "rejected gossip messages");
+    createDirectory(this.directory.resolve(INVALID_BLOCK_DIR), INVALID_BLOCK_DIR, "invalid blocks");
+  }
+
+  public void saveGossipMessageDecodingError(
+      final String topic, final Optional<UInt64> arrivalTimestamp, final Bytes originalMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String formattedTimestamp = formatOptionalTimestamp(arrivalTimestamp);
+    final String fileName = String.format("%s.ssz", formattedTimestamp);
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR)
+            .resolve(DECODING_ERROR_SUB_DIR)
+            .resolve(topic.replaceAll("/", "_"));
+    final String identifiers = String.format("on topic %s at %s", topic, formattedTimestamp);
+    saveBytesToFile(
+        "gossip message with decoding error",
+        identifiers,
+        topicPath.resolve(fileName),
+        originalMessage);
+  }
+
+  public void saveGossipRejectedMessageToFile(
+      final String topic, final Optional<UInt64> arrivalTimestamp, final Bytes decodedMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String formattedTimestamp = formatOptionalTimestamp(arrivalTimestamp);
+    final String fileName = String.format("%s.ssz", formattedTimestamp);
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR).resolve(REJECTED_SUB_DIR).resolve(topic.replaceAll("/", "_"));
+    final String identifiers = String.format("on topic %s at %s", topic, formattedTimestamp);
+    saveBytesToFile(
+        "rejected gossip message", identifiers, topicPath.resolve(fileName), decodedMessage);
+  }
+
+  public void saveInvalidBlockToFile(
+      final UInt64 slot, final Bytes32 blockRoot, final Bytes blockSsz) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s_%s.ssz", slot, blockRoot.toUnprefixedHexString());
+    final String identifiers = String.format("at slot %s(%s)", slot, blockRoot);
+    saveBytesToFile(
+        "invalid block", identifiers, Path.of(INVALID_BLOCK_DIR).resolve(fileName), blockSsz);
```

Comment:
the more i htink about it, the more i think the message previously was correct including the level, and this change we just dont have the 'full block data' at the end.
I suggest again that we fix this logging so that its just less painful than passing it to the file saving and confusing concerns. I think the previous warn of `Rejecting invalid block at slot 1410696 with root 0xd6107c0e5eeaecd1ba385e65336cbf3c043110a2a399497ebb873229c57dcdc4 because FAILED_STATE_TRANSITION.` was about perfect, we just remove the 'full block data:' and write it to file ,and then its much easier to see old v. new functionality

Reply:
Okay will update it. Just keeping in mind that a user can have both the `--Xlog-include-p2p-warnings-enabled` and the new `--Xp2p-dumps-to-file-enabled` flag on at the same time. In that case there would be two logs in a row of the same thing but one would not have the full block data. However, it is unlikely a user would do that and that is their choice.

Reply:
point being then we can compare both, so having the same message or just 1 at same level simplifies seeing the data